authres/parse.go

```diff
@@ -293,10 +331,114 @@ func parseResult(s string) (Result, error) {
 	return r, nil
 }
 
-func parseParam(s string) (k string, v string, err error) {
-	kv := strings.SplitN(s, "=", 2)
-	if len(kv) != 2 {
-		return "", "", errors.New("msgauth: malformed authentication method and value")
+// keyValue parses a sequence of key=value parameters
+func (p *parser) keyValue() (k, v string, err error) {
+	k, err = p.readKey()
+	if err != nil {
+		return
+	}
+	v, err = p.readValue()
+	if err != nil {
+		return
 	}
-	return strings.ToLower(strings.TrimSpace(kv[0])), strings.TrimSpace(kv[1]), nil
+	return
+}
+
+// readKey reads a method, reason or ptype.property as defined in RFC 8601
+// Section 2.2. Ignore the method-version of the methodspec. Stop at EOF or the
+// equal sign.
+func (p *parser) readKey() (k string, err error) {
+	var c byte
+	for err != io.EOF {
+		c, err = p.r.ReadByte()
+		if err != nil {
+			break
+		}
+		switch c {
+		case ';':
+			err = io.EOF
+			break
+		case '=':
+			break
+		case '(':
+			p.r.UnreadByte()
+			_, err = p.readComment()
+			continue
+		case '/':
+			p.r.ReadBytes('=')
+			p.r.UnreadByte()
+		default:
+			if !unicode.IsSpace(rune(c)) {
+				k += string(c)
+			}
+		}
+		if c == '=' {
+			break
+		}
+	}
+	k = strings.TrimSpace(strings.ToLower(k))
+	return
+}
+
+// readValue reads a result or value as defined in RFC 8601 Section 2.2. Value
+// is defined as either a token or quoted string according to RFC 2045 Section
+// 5.1. Stop at EOF, white space or semi-colons.
+func (p *parser) readValue() (v string, err error) {
+	var c byte
+	for err != io.EOF {
+		c, err = p.r.ReadByte()
+		if err != nil {
+			break
+		}
+		switch c {
+		case ';':
+			err = io.EOF
+			break
+		case '(':
+			p.r.UnreadByte()
+			_, err = p.readComment()
+			continue
+		case '"':
+			v, err = p.r.ReadString(c)
+			v = strings.TrimSuffix(v, string(c))
+		default:
+			if !unicode.IsSpace(rune(c)) {
+				v += string(c)
+			}
+		}
+		if unicode.IsSpace(rune(c)) {
+			if v != "" {
+				break
+			}
+		}
+	}
+	v = strings.TrimSpace(v)
+	return
+}
+
+func (p *parser) readComment() (comment string, err error) {
+	count := 0
+	var c byte
+	for {
+		c, err = p.r.ReadByte()
+		if err != nil {
+			break
+		}
+		switch c {
+		case '\\':
+			c, _ = p.r.ReadByte()
+			comment += "\\" + string(c)
+		case '(':
```

Comment:
the `CFWS` referenced by https://datatracker.ietf.org/doc/html/rfc8601#section-2.2 is defined in https://www.rfc-editor.org/rfc/rfc5322#section-3.2.2 (internet message / email):

```
FWS             =   ([*WSP CRLF] 1*WSP) /  obs-FWS
                                       ; Folding white space

ctext           =   %d33-39 /          ; Printable US-ASCII
                    %d42-91 /          ;  characters not including
                    %d93-126 /         ;  "(", ")", or "\"
                    obs-ctext

ccontent        =   ctext / quoted-pair / comment

comment         =   "(" *([FWS] ccontent) [FWS] ")"

CFWS            =   (1*([FWS] comment) [FWS]) / FWS
```

this code is not entirely future-proof, though right now, i welcome everything that helps me handle my emails..
even [net/mail](https://pkg.go.dev/net/mail#pkg-overview) doesn't go the whole way:

> The full range of spacing (the CFWS syntax element) is not supported, such as breaking addresses across lines.

how about including the parentheses in the output string (doing `comment += string(c)` for both `(` and `)`) and leaving parsing of comments up to other or future implementations?
in that case, the backslashes should also be kept.

i can think of invalid comments that would be accepted by this parser. particularly, the restrictions for `ctext`.
the implementation before this pr was more restrictive than the grammar. this replacement is more liberal.
that change could be understood as a changing interface.